src/snowcli/app/cli_app.py

```diff
@@ -1,48 +1,86 @@
 from __future__ import annotations
 
-import importlib
 import logging
 from pathlib import Path
+from typing import Optional
 
-import typer
 import click
+import typer
 
 from snowcli import __about__
+from snowcli.app.commands_registration.commands_registration_with_callbacks import (
+    CommandsRegistrationWithCallbacks,
+)
 from snowcli.app.dev.commands_structure import generate_commands_structure
+from snowcli.app.dev.docs.generator import generate_docs
+from snowcli.app.dev.pycharm_remote_debug import (
+    setup_pycharm_remote_debugger_if_provided,
+)
 from snowcli.app.main_typer import SnowCliMainTyper
 from snowcli.config import config_init, cli_config
-from snowcli.app.dev.docs.generator import generate_docs
 from snowcli.output.formats import OutputFormat
 from snowcli.output.printing import print_result
 from snowcli.output.types import CollectionResult
-from snowcli.app.dev.pycharm_remote_debug import (
-    setup_pycharm_remote_debugger_if_provided,
-)
 
 app: SnowCliMainTyper = SnowCliMainTyper()
 log = logging.getLogger(__name__)
 
+_commands_registration = CommandsRegistrationWithCallbacks()
+
+
+def _do_not_execute_on_completion(callback):
+    def enriched_callback(value):
+        if click.get_current_context().resilient_parsing:
+            return
+        callback(value)
+
+    return enriched_callback
+
+
+def _commands_registration_callback(value: bool):
+    if value:
+        _commands_registration.register_commands_if_ready_and_not_registered_yet()
+    # required to make the tests working
+    _commands_registration.reset_running_instance_registration_state()
```

Comment:
I don't have a better idea how to do it better.

Main `app` instance seems to be reused in tests (for sure in parameterized tests), but command needs to be registered in each execution of this `app`.
Required callbacks gathering is done only once (during the import of the module) while commands registration is done multiple times (for every runner.invoke)

Reply:
Can we rebuild test instance every time? Or that won't solve the problem?

Reply:
Currently we create a new `runner` instance from fixture for each test (`scope=function`).
```python
@pytest.fixture(scope="function")
def runner(test_snowcli_config):
    from snowcli.app.cli_app import app
    return SnowCLIRunner(app, test_snowcli_config)
```
However imported `app` is cached.
I was able to do resolve caching issue with following code:
```python
@pytest.fixture(scope="function")
def runner(test_snowcli_config):
    from snowcli.app.cli_app import app
    yield SnowCLIRunner(app, test_snowcli_config)
    sys.modules.pop("snowcli.app.cli_app")
```
But still there are tests which do multiple calls to runner.invoke in the same test case and for such tests removing imported modules won't help.

I still don't see a better option than `reset_running_instance_registration_state`.

Reply:
I made the comment more verbose.

Reply:
How about reloading the module or invalidating cache?
https://docs.python.org/3/library/importlib.html#importlib.reload